File: app/src/main/java/com/novel/read/widget/dialog/ReadSettingDialog.java
package com.novel.read.widget.dialog;

import android.app.Activity;
import android.app.Dialog;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.novel.read.R;
import com.novel.read.adapter.PageStyleAdapter;
import com.novel.read.widget.page.PageLoader;
import com.novel.read.widget.page.PageMode;
import com.novel.read.widget.page.PageStyle;
import com.novel.read.widget.page.ReadSettingManager;

import java.util.Arrays;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ReadSettingDialog extends Dialog {

    private static final String TAG = "ReadSettingDialog";

    @BindView(R.id.tv_simple)
    TextView mTvSimple;
    @BindView(R.id.tv_trans)
    TextView mTvTrans;

    @BindView(R.id.read_setting_tv_font_minus)
    ImageView mTvFontMinus;

    @BindView(R.id.read_setting_tv_font_plus)
    ImageView mTvFontPlus;

    @BindView(R.id.read_setting_rv_bg)
    RecyclerView mRvBg;

    /************************************/
    private PageStyleAdapter mPageStyleAdapter;
    private ReadSettingManager mSettingManager;
    private PageLoader mPageLoader;
    private Activity mActivity;

    private PageMode mPageMode;
    private PageStyle mPageStyle;

    private int mBrightness;
    private int mTextSize;

    private boolean isBrightnessAuto;
    private boolean isTextDefault;

    private int convertType;

    public ReadSettingDialog(@NonNull Activity activity, PageLoader mPageLoader) {
        super(activity, R.style.ReadSettingDialog);
        mActivity = activity;
        this.mPageLoader = mPageLoader;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.layout_setting);
        ButterKnife.bind(this);
        setUpWindow();
        initData();
        initWidget();
        initClick();
    }

    //设置Dialog显示的位置
    private void setUpWindow() {
        Window window = getWindow();
        WindowManager.LayoutParams lp = window.getAttributes();
        lp.width = WindowManager.LayoutParams.MATCH_PARENT;
        lp.height = WindowManager.LayoutParams.WRAP_CONTENT;
        lp.gravity = Gravity.BOTTOM;
        window.setAttributes(lp);
    }

    private void initData() {
        mSettingManager = ReadSettingManager.getInstance();

        isBrightnessAuto = mSettingManager.isBrightnessAuto();
        mBrightness = mSettingManager.getBrightness();
        mTextSize = mSettingManager.getTextSize();
        isTextDefault = mSettingManager.isDefaultTextSize();
        mPageMode = mSettingManager.getPageMode();
        mPageStyle = mSettingManager.getPageStyle();
        convertType = mSettingManager.getConvertType();
        if (convertType == 0) {
            mTvSimple.setSelected(true);
            mTvTrans.setSelected(false);
        } else {
            mTvSimple.setSelected(false);
            mTvTrans.setSelected(true);
        }
    }

    private void initWidget() {
        //RecyclerView
        setUpAdapter();
    }

    private void setUpAdapter() {
        Drawable[] drawables = {
                getDrawable(R.color.nb_read_bg_1)
                , getDrawable(R.color.nb_read_bg_2)
                , getDrawable(R.color.nb_read_bg_4)
                , getDrawable(R.color.nb_read_bg_5)};

        mPageStyleAdapter = new PageStyleAdapter(Arrays.asList(drawables), mPageLoader);
        mRvBg.setLayoutManager(new GridLayoutManager(getContext(), 5));
        mRvBg.setAdapter(mPageStyleAdapter);

        mPageStyleAdapter.setPageStyleChecked(mPageStyle);

    }


    private Drawable getDrawable(int drawRes) {
        return ContextCompat.getDrawable(getContext(), drawRes);
    }

    private void initClick() {

        //字体大小调节
        mTvFontMinus.setOnClickListener(
                (v) -> {
                    int fontSize = mSettingManager.getTextSize() - 1;
                    if (fontSize < 0) return;
                    mPageLoader.setTextSize(fontSize);
                }
        );

        mTvFontPlus.setOnClickListener(
                (v) -> {
                    int fontSize = mSettingManager.getTextSize() + 1;
                    mPageLoader.setTextSize(fontSize);
                }
        );

        mTvSimple.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (convertType == 0) {
                    return;
                }
                mTvSimple.setSelected(true);
                mTvTrans.setSelected(false);
                mSettingManager.setConvertType(0);
                convertType = 0;
                mPageLoader.setTextSize(mSettingManager.getTextSize());
            }
        });

        mTvTrans.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (convertType == 1) {
                    return;
                }
                mTvSimple.setSelected(false);
                mTvTrans.setSelected(true);
                mSettingManager.setConvertType(1);
                convertType = 1;
                mPageLoader.setTextSize(mSettingManager.getTextSize());
            }
        });

    }

}

File: app/src/main/java/com/novel/read/widget/dialog/ReadSettingDialog.kt
package com.novel.read.widget.dialog

import android.app.Activity
import android.app.Dialog
import android.graphics.drawable.Drawable
import android.view.Gravity
import android.view.View
import android.view.WindowManager
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import com.novel.read.R
import com.novel.read.adapter.PageStyleAdapter
import com.novel.read.utlis.ScreenUtils
import com.novel.read.widget.page.PageLoader
import com.novel.read.widget.page.PageMode
import com.novel.read.widget.page.PageStyle
import com.novel.read.widget.page.ReadSettingManager
import kotlinx.android.synthetic.main.layout_setting.*

class ReadSettingDialog(mActivity: Activity, private var mPageLoader: PageLoader) :
    Dialog(mActivity, R.style.ReadSettingDialog) {

    private lateinit var mPageStyleAdapter: PageStyleAdapter
    private var mSettingManager: ReadSettingManager = ReadSettingManager.getInstance()

    private var mPageMode: PageMode? = null
    private var mPageStyle: PageStyle? = null

    private var mBrightness: Int = 0
    private var mTextSize: Int = 0

    private var isBrightnessAuto: Boolean = false
    private var isTextDefault: Boolean = false

    private var convertType: Int = 0

    init {
        setContentView(R.layout.layout_setting)
        setUpWindow()
        initData()
        initWidget()
        initClick()
        initPageMode()
    }

    //设置Dialog显示的位置
    private fun setUpWindow() {
        val window = window
        val lp = window!!.attributes
        lp.width = WindowManager.LayoutParams.MATCH_PARENT
        lp.height = WindowManager.LayoutParams.WRAP_CONTENT
        lp.gravity = Gravity.BOTTOM
        window.attributes = lp
    }

    private fun initData() {

        isBrightnessAuto = mSettingManager.isBrightnessAuto
        mBrightness = mSettingManager.brightness
        mTextSize = mSettingManager.textSize
        isTextDefault = mSettingManager.isDefaultTextSize
        mPageMode = mSettingManager.pageMode
        mPageStyle = mSettingManager.pageStyle
        convertType = mSettingManager.convertType
        if (convertType == 0) {
            tv_simple.isSelected = true
            tv_trans.isSelected = false
        } else {
            tv_simple.isSelected = false
            tv_trans.isSelected = true
        }
        tv_size.text = "$mTextSize"
    }

    private fun initWidget() {
        setUpAdapter()
    }

    private fun setUpAdapter() {
        val drawables = arrayOf(
            getDrawable(R.color.read_bg_one),
            getDrawable(R.color.read_bg_two),
            getDrawable(R.color.read_bg_four),
            getDrawable(R.color.read_bg_five)
        )

        mPageStyleAdapter = PageStyleAdapter(listOf(*drawables), mPageLoader)
        read_setting_rv_bg.layoutManager = GridLayoutManager(context, 4)
        read_setting_rv_bg.adapter = mPageStyleAdapter

        mPageStyleAdapter.setPageStyleChecked(mPageStyle!!)

    }

    private fun getDrawable(drawRes: Int): Drawable? {
        return ContextCompat.getDrawable(context, drawRes)
    }

    private fun initClick() {

        //字体大小调节
        read_setting_tv_font_minus.setOnClickListener {
            val fontSize = mSettingManager.textSize - 1
            if (fontSize < 0) {
                return@setOnClickListener
            }
            mPageLoader.setTextSize(fontSize)
            tv_size.text = "$fontSize"
        }

        read_setting_tv_font_plus.setOnClickListener {
            val fontSize = mSettingManager.textSize + 1
            mPageLoader.setTextSize(fontSize)
            tv_size.text = "$fontSize"
        }

        tv_font_default.setOnClickListener {
            val fontSize = ScreenUtils.spToPx(16)
            mSettingManager.textSize = fontSize
            mPageLoader.setTextSize(fontSize)
            tv_size.text = "$fontSize"
        }

        tv_simple.setOnClickListener(View.OnClickListener {
            if (convertType == 0) {
                return@OnClickListener
            }
            tv_simple.isSelected = true
            tv_trans.isSelected = false
            mSettingManager.convertType = 0
            convertType = 0
            mPageLoader.setTextSize(mSettingManager.textSize)
        })

        tv_trans.setOnClickListener(View.OnClickListener {
            if (convertType == 1) {
                return@OnClickListener
            }
            tv_simple.isSelected = false
            tv_trans.isSelected = true
            mSettingManager.convertType = 1
            convertType = 1
            mPageLoader.setTextSize(mSettingManager.textSize)
        })

        //Page Mode 切换
        read_setting_rg_page_mode.setOnCheckedChangeListener { group, checkedId ->
            val pageMode: PageMode = when (checkedId) {
                R.id.read_setting_rb_simulation -> PageMode.SIMULATION
                R.id.read_setting_rb_cover -> PageMode.COVER
                R.id.read_setting_rb_scroll -> PageMode.SCROLL
                R.id.read_setting_rb_none -> PageMode.NONE
                else -> PageMode.SIMULATION
            }
            mPageLoader.setPageMode(pageMode)
        }

    }

    private fun initPageMode() {
        when (mPageMode) {
            PageMode.SIMULATION -> read_setting_rb_simulation.isChecked = true
            PageMode.COVER -> read_setting_rb_cover.isChecked = true
            PageMode.NONE -> read_setting_rb_none.isChecked = true
            PageMode. SCROLL -> read_setting_rb_scroll.isChecked = true
            else -> {

            }
        }
    }

    companion object {
        private val TAG = "ReadSettingDialog"
    }

}
